csrc/multidevice/cuda_p2p.cpp

```diff
@@ -0,0 +1,70 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <cuda_utils.h>
+#include <multidevice/cuda_p2p.h>
+
+namespace nvfuser {
+
+namespace getZcopy {
+
+void RecvPost(const P2pIpcHandle& ipc_handles, int64_t count, CUstream stream) {
+  // wait for sender to be ready
+  NVFUSER_CUDA_SAFE_CALL(cuStreamWaitValue32(
+      stream,
+      reinterpret_cast<CUdeviceptr>(ipc_handles.local().semaphore()),
+      (cuuint32_t)(IpcSemaphore::kInUse),
+      CU_STREAM_WAIT_VALUE_EQ));
+  // RDMA get the data from the sender
+  NVFUSER_CUDA_RT_SAFE_CALL(cudaMemcpyAsync(
+      ipc_handles.local().ptr(),
+      ipc_handles.peer().ptr(),
+      count,
+      cudaMemcpyDeviceToDevice,
+      stream));
+  // Signals completion to self
```

Comment:
Can you explain why this is needed in addition to the other completion signal below?

Reply:
This step is needed to reset the semaphore in the case of a future reuse.

see also https://github.com/NVIDIA/Fuser/pull/3911#discussion_r1983280759

Reply:
Got it -- both semaphores need to have the same value. This way, waiting for local is equivalent to waiting for peer.